finetune/base_models/oscar/featurizer.py

```diff
@@ -0,0 +1,261 @@
+import tensorflow as tf
+from finetune.base_models.gpt.featurizer import dropout, embed, split_heads, merge_heads
+from finetune.util.shapes import shape_list
+from finetune.base_models.oscar.ra import recursive_agg
+from finetune.optimizers.recompute_grads import recompute_grad
+from finetune.nn.add_auxiliary import add_auxiliary
+
+import functools
+
+
+def cast_maybe(t, dtype):
+    if t.dtype == dtype:
+        return t
+    return tf.cast(t, dtype=dtype)
+
+
+def norm(x, scope, axis=-1, e=None, fp16=False, debug=False):
+    with tf.variable_scope(scope):
+        e = e or 1e-5
+        n_state = shape_list(x)[-1]
+        g = tf.get_variable("g", [n_state], initializer=tf.constant_initializer(1))
+        b = tf.get_variable("b", [n_state], initializer=tf.constant_initializer(0))
+        if fp16:
+            g = tf.cast(g, tf.float16)
+            b = tf.cast(b, tf.float16)
+        u = tf.reduce_mean(x, axis=axis, keepdims=True)
+        s = tf.reduce_mean(tf.square(x - u), axis=axis, keepdims=True)
+        if debug:
+            u = tf.Print(u, [u[0, 0], s[0, 0]])
+        x = (x - u) * tf.rsqrt(s + e)
+        x = x * g + b
+        return x
+
+
+def embed_no_timing(X, we):
+    return tf.gather(we, X[:, :, 0])
+
+
+def time_to_batch(value, dilation, pad_with=0):
+    with tf.name_scope('time_to_batch'):
+        shape = tf.shape(value)
+        pad_elements = dilation - 1 - (shape[1] + dilation - 1) % dilation
```

Comment:
`(shape[1] + dilation -1)` % dilation should be the same as `(shape[1] - 1) % dilation`

Reply:
And the ones should cancel as well?  So this just becomes `dilation - shape[1] % dilation`?

Reply:
Maybe I'm missing some edge case though

Reply:
dilation = 2 and shape[1] = 4
the original does 2 - 1 - (4 + 2 - 1) % 2 = 1 - 5 % 2 = 0
the first revision does 2 - 1 - (4 - 1) % 2 = 1 - 3 % 2 = 0  ( I agree with this change entirely, thanks)
second revision does 2 - 4 % 2 = 2 - 0 = 2 (I disagree with this change)